### PR TITLE
Feat/credentials

### PR DESCRIPTION
**Pull request by noandrea**

### CODE REVIEW

File: x/credential/README.md [hunk 1]
```diff
@@ -0,0 +1,18 @@
+# DID Module
```
Comment (gsora) [inline x/credential/README.md]: Should this README stay here?

File: x/credential/client/cli/query.go [hunk 2]
```diff
@@ -0,0 +1,158 @@
...
+	return cmd
+}
+
+func NewQueryRevocationListCmd() *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "revocation-list [id]",
+		Short: "get the revocation list for a credential issuer",
+		Args:  cobra.ExactArgs(1),
+		RunE: func(cmd *cobra.Command, args []string) error {
+			//clientCtx, err := client.GetClientQueryContext(cmd)
```
Comment (gsora) [inline x/credential/client/cli/query.go]: Should this command stay or go away until we're ready for a full implementation?

File: x/credential/credential.go [hunk 3]
```diff
@@ -0,0 +1,151 @@
...
+		return nil, err
+	}
+	wc.PublicVerifiableCredential.CredentialSubject = sbj
+	return wc.PublicVerifiableCredential, nil
+}
+
+func (wc *WrappedCredential) GetBytes() []byte {
+	dAtA, err := json.Marshal(wc)
+	if err != nil {
+		panic(err) //[(gogoproto.sizer) = true, (gogoproto.marshaler) = true,  (gogoproto.unmarshaler) = true];
```
Comment (gsora) [inline x/credential/credential.go:131]: Is a panic really needed here? Can't a good old error return work just as well?

File: x/credential/credential.go [hunk 4]
```diff
@@ -0,0 +1,151 @@
...
+
+func (pvc PublicVerifiableCredential) GetIssuerDID() did.DID {
+	return did.DID(pvc.Issuer)
+}
+
+// GetBytes is a helper for serializing
+func (pvc PublicVerifiableCredential) GetBytes() []byte {
+	dAtA, err := pvc.Marshal()
+	if err != nil {
+		panic(err) //[(gogoproto.sizer) = true, (gogoproto.marshaler) = true,  (gogoproto.unmarshaler) = true];
```
Comment (gsora) [inline x/credential/credential.go:184]: Here as well.

File: x/credential/credential.go [hunk 5]
```diff
@@ -0,0 +1,151 @@
...
+	return dAtA
+}
+
+// Validate validates a verifiable credential against a provided public key
+func (wc WrappedCredential) Validate(
+	pk cryptotypes.PubKey,
+) bool {
+	s, err := base64.StdEncoding.DecodeString(wc.Proof.Signature)
+	if err != nil {
+		panic(err)
```
Comment (gsora) [inline x/credential/credential.go]: And here.

File: x/credential/errors.go [hunk 6]
```diff
@@ -0,0 +1,25 @@
+package credential
+
+// DONTCOVER
+
+import (
+	sdkerrors "github.com/cosmos/cosmos-sdk/types/errors"
+)
+
+// x/did module sentinel errors
```
Comment (gsora) [inline x/credential/errors.go]: Wrong package name here?

File: x/credential/errors.go [hunk 7]
```diff
@@ -0,0 +1,25 @@
...
+var (
+	ErrCredentialDefinitionFound     = sdkerrors.Register(ModuleName, 2100, "credential definition found")
+	ErrCredentialDefinitionNotFound  = sdkerrors.Register(ModuleName, 2101, "credential definition not found")
+	ErrCredentialDefinitionCorrupted = sdkerrors.Register(ModuleName, 2109, "credential definition corrupted")
+
+	ErrVerifiableCredentialNotFound = sdkerrors.Register(ModuleName, 2102, "vc not found")
+	ErrVerifiableCredentialFound    = sdkerrors.Register(ModuleName, 2103, "vc found")
+	ErrDidDocumentDoesNotExist      = sdkerrors.Register(ModuleName, 2104, "did does not exist in the store")
+	ErrVerifiableCredentialIssuer   = sdkerrors.Register(ModuleName, 2105, "provided verifiable credential and did public key do not match")
+	ErrMessageSigner                = sdkerrors.Register(ModuleName, 2106, "message signer does not match provided did")
```
Comment (gsora) [inline x/credential/errors.go:20]: Is this error used to check against dids?

File: x/credential/keeper/credential.go [hunk 8]
```diff
@@ -0,0 +1,72 @@
...
+
+// GetPublicCredential retrieve a public verifiable credential by its key.
+// The boolean return will be false if the credential is not found
+func (k Keeper) GetPublicCredential(ctx sdk.Context, key string) (credential.PublicVerifiableCredential, bool) {
+	val, found := k.Get(ctx, []byte(key), credential.PublicCredentialKey, func(value []byte) (interface{}, bool) {
+		var data credential.PublicVerifiableCredential
+		ok := k.Unmarshal(value, &data)
+		return data, ok
+	})
+	return val.(credential.PublicVerifiableCredential), found
```
Comment (gsora) [inline x/credential/keeper/credential.go:39]: Could this type assertion ever fail?

File: x/credential/keeper/credential.go [hunk 9]
```diff
@@ -0,0 +1,72 @@
...
+	var pvcs []*credential.PublicVerifiableCredential
+
+	iterator := k.GetAll(ctx, credential.PublicCredentialKey)
+	defer iterator.Close()
+
+	for ; iterator.Valid(); iterator.Next() {
+		var pvc credential.PublicVerifiableCredential
+		err := k.cdc.Unmarshal(iterator.Value(), &pvc)
+		if err != nil {
+			panic(err)
```
Comment (gsora) [inline x/credential/keeper/credential.go:65]: Is a panic here needed?

File: x/credential/msg_validation.go [hunk 10]
```diff
@@ -0,0 +1,35 @@
+package credential
+
+import (
+	"fmt"
+
+	"github.com/elesto-dao/elesto/x/did"
+)
+
+func (msg MsgPublishCredentialDefinitionRequest) ValidateBasic() error {
+	if !did.IsValidDID(msg.CredentialDefinition.Id) {
```
Comment (gsora) [inline x/credential/msg_validation.go:14]: `msg.CredentialDefinition` is a pointer, could be nil: should we check for this case?

File: x/credential/simulation/operations.go [hunk 11]
```diff
@@ -0,0 +1,568 @@
+package simulation
+
+//
```
Comment (gsora) [inline x/credential/simulation/operations.go:3]: Should this simulation be commented entirely?

**Review by gsora — COMMENTED** [review]
Left some comments, thanks for submitting this PR!

**Review by PaddyMc — APPROVED** [review]
Ace, gg!!! :mage: 

I think we should merge https://github.com/elesto-dao/elesto/pull/113 into this branch then this branch into `main` ASAP, it's defo in a good enough state to be merged :tada:

File: x/credential/client/cli/query.go (hunk 2 above)
Comment (noandrea) [reply x/credential/client/cli/query.go]: this has been addressed in #113

File: x/credential/credential.go (hunk 5 above)
Comment (noandrea) [reply x/credential/credential.go]: Addressed in #113

File: x/credential/msg_validation.go (hunk 10 above)
Comment (noandrea) [reply x/credential/msg_validation.go:14]: good catch! fixed in #113

File: x/credential/errors.go (hunk 6 above)
Comment (noandrea) [reply x/credential/errors.go]: fixed in #113

File: x/credential/simulation/operations.go (hunk 11 above)
Comment (noandrea) [reply x/credential/simulation/operations.go:3]: I'll leave it like that until we implement it

File: x/credential/README.md (hunk 1 above)
Comment (noandrea) [reply x/credential/README.md]: fixed in #113

File: x/credential/credential.go (hunk 3 above)
Comment (noandrea) [reply x/credential/credential.go:131]: tracked in #114

File: x/credential/credential.go (hunk 4 above)
Comment (noandrea) [reply x/credential/credential.go:184]: tracked in #114

File: x/credential/errors.go (hunk 7 above)
Comment (noandrea) [reply x/credential/errors.go:20]: tracked in #114

File: x/credential/keeper/credential.go (hunk 8 above)
Comment (noandrea) [reply x/credential/keeper/credential.go:39]: tracked in #114

File: x/credential/keeper/credential.go (hunk 9 above)
Comment (noandrea) [reply x/credential/keeper/credential.go:65]: tracked in #114

**Pull request merged**